sdk/go/pulumi-language-go/main.go

```diff
@@ -135,18 +137,50 @@ func (host *goLanguageHost) Run(ctx context.Context, req *pulumirpc.RunRequest)
 		return nil, errors.Wrap(err, "failed to prepare environment")
 	}
 
+	// by default we try to run a named executable on the path, but we will fallback to 'go run' style execution
+	goRunInvoke := false
+
 	// The program to execute is simply the name of the project.  This ensures good Go toolability, whereby
 	// you can simply run `go install .` to build a Pulumi program prior to running it, among other benefits.
 	program, err := findProgram(req.GetProject())
 	if err != nil {
-		return nil, errors.Wrap(err, "problem executing program (could not run language executor)")
+		const message = "problem executing program (could not run language executor)"
+		if err.Error() == fmt.Sprintf(unableToFindProgramTemplate, req.GetProject()) {
+			logging.V(5).Infof("Unable to find program %s in $PATH, attempting invocation via 'go run'", program)
+			program, err = findProgram("go")
+			if err != nil {
+				return nil, errors.Wrap(err, message)
+			}
+			goRunInvoke = true
+		} else {
+			return nil, errors.Wrap(err, message)
+		}
```

Comment:
Is this same technique something we can do for .NET as well to accomplish the opposite?

Would love to have a consistent approach here if possible.

Reply:
Great idea, can you take a look at https://github.com/pulumi/pulumi/issues/3509  and verify that I've captured your thoughts accurately. 

Happy to pull this onto my plate if we think it's an immediate priority.